staging/kos/pkg/controllers/connectionagent/connectionagent.go

```diff
@@ -0,0 +1,1059 @@
+	if !newIfcNeedsToBeCreated {
+		// If we're here the new and the old interface are equals, so we return
+		return nil
+	}
+
+	var err error
+	if oldIfcExists {
+		// If we're here there's an old interface which differs from the new one.
+		// We delete the old interface before creating a new one.
+		err = ca.netFabric.DeleteRemoteIfc(oldIfc)
+		if err != nil {
+			ca.testAndSwapRemoteAttIfcIfVNIsMatch(attNSN, oldIfc)
+			return err
+		}
+	}
+	// create the new interface
+	err = ca.netFabric.CreateRemoteIfc(newIfc)
+	if err != nil {
+		ca.removeRemoteAttIfc(attNSN)
+	}
+	return err
+}
+
+func (ca *ConnectionAgent) attemptToDeleteLocalAttIfc(nsn k8stypes.NamespacedName, vni uint32) (string, error) {
+	ifc, ifcFound := ca.getLocalAttIfc(nsn)
+	if ifcFound && ifc.VNI == vni {
+		err := ca.netFabric.DeleteLocalIfc(ifc)
+		if err == nil {
+			ca.removeLocalAttIfc(nsn)
+			return ifc.Name, nil
+		}
+		return "", err
+	}
+	return "", nil
+}
+
+func (ca *ConnectionAgent) attemptToDeleteRemoteAttIfc(nsn k8stypes.NamespacedName, vni uint32) error {
+	ifc, ifcFound := ca.getRemoteAttIfc(nsn)
+	if ifcFound && ifc.VNI == vni {
+		err := ca.netFabric.DeleteRemoteIfc(ifc)
+		if err == nil {
+			ca.removeRemoteAttIfc(nsn)
+		}
+		return err
+	}
+	return nil
+}
+
+func (ca *ConnectionAgent) updateVNStateForExistingLocalAtt(localAtt *netv1a1.NetworkAttachment) (firstInVN bool) {
+	vni := localAtt.Spec.VNI
+
+	ca.vniToVnStateMutex.Lock()
+	defer ca.vniToVnStateMutex.Unlock()
+
+	localAttVnState := ca.vniToVnState[vni]
+
+	if localAttVnState == nil {
+		remoteAttsInformer, remoteAttsLister := v1a1AttsCustomNamespaceInformerAndLister(ca.kcs,
+			resyncPeriod,
+			localAtt.Namespace,
+			fromFieldsSelectorToTweakListOptionsFunc(ca.remoteAttInVNWithVirtualIPHostIPAndIfcSelector(vni)))
+
+		remoteAttsInformer.AddIndexers(map[string]k8scache.IndexFunc{attMACIndexName: attachmentMACAddr})
+
+		remoteAttsInformer.AddEventHandler(k8scache.ResourceEventHandlerFuncs{
+			AddFunc:    ca.onRemoteAttAdded,
+			UpdateFunc: ca.onRemoteAttUpdated,
+			DeleteFunc: ca.onRemoteAttRemoved,
+		})
+
+		remoteAttsInformerStopCh := make(chan struct{})
+		go remoteAttsInformer.Run(aggregateTwoStopChannels(ca.stopCh, remoteAttsInformerStopCh))
+
+		localAttVnState = &vnState{
+			remoteAttsInformer:       remoteAttsInformer,
+			remoteAttsInformerStopCh: remoteAttsInformerStopCh,
+			remoteAttsLister:         remoteAttsLister,
+			namespace:                localAtt.Namespace,
+			localAtts:                make(map[string]struct{}, 1),
+		}
+		ca.vniToVnState[vni] = localAttVnState
+		firstInVN = true
+	}
+
+	localAttVnState.localAtts[localAtt.Name] = struct{}{}
+	return
+}
+
+func (ca *ConnectionAgent) addRemoteAttToVNState(remAttName string, vni uint32) (added bool) {
+	ca.vniToVnStateMutex.Lock()
+	defer ca.vniToVnStateMutex.Unlock()
+	vnState := ca.vniToVnState[vni]
+	if vnState != nil {
+		if vnState.remoteAtts == nil {
```

Comment:
The code will be simpler if we initialize the `vnState.remoteAtts` when `vnState` is first constructed.

Reply:
Suggestion implemented in a597f1e.